### PR TITLE
Changes default behavior for DynamoDb Enhanced atomic counter extension

### PR DESCRIPTION
**Pull request by cenedhryn**
## Motivation and Context
The initial implementation of the atomic counter extension made the assumption that it was not necessary to modify the item itself while adding an update expression, because users should not include attributes that were tagged as an atomic counter in the item to be updated and if they did, they would get an error back. 

For instance, if attribute `accessCounter` of the class `Balance` was marked with the `AtomicCounter` attribute, this would result in an error:
~~~
Balance balanceToUpdate = new Balance();
balanceToUpdate.accessCounter(4);
table.updateItem(balanceToUpdate);
~~~
Which makes sense since `accessCounter` will be updated anyway in the `AtomicCounterExtension`, and the assumption was that an error is better than silently changing the item when `updateItem` was called. This assumption did not take into account that a very common, and natural, behavior for users is read-modify-update:
~~~
Balance balanceToUpdate = table.getItem(key);
balanceToUpdate.anImportantAttribute(newValue);
table.updateItem(balanceToUpdate);
~~~
Here, the user will actually get the same error as above, because while they did not explicitly set `accessCounter`, the instance returned from DDB will have it set, and it's included when the user calls `updateItem`. 

Therefore, we should change this behavior to filter out any counter attributes in the item map if they're present when we add an update expression for the atomic counter. From a backwards compatibility perspective, this change should have little effect on existing user code. 

## Modifications
AtomicCounterExpression filters any attributes from the item that are marked as atomic counters. If it finds any, it also logs a debug level warning to provide context to users.  

## Testing
Unit and functional testing

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AtomicCounterTest.java [hunk 1]
```diff
@@ -112,15 +112,31 @@ public void createViaPut_incrementsCorrectly() {
...
+    public void updateItem_retrievedFromDb_shouldNotThrowException() {
+        AtomicCounterRecord record = new AtomicCounterRecord();
+        record.setId(RECORD_ID);
+        record.setAttribute1(STRING_VALUE);
+        mappedTable.updateItem(record);
+
+        AtomicCounterRecord retrievedRecord = mappedTable.getItem(record);
+        retrievedRecord.setAttribute1("ChangingThisAttribute");
+
+        assertThat(mappedTable.updateItem(retrievedRecord)).isNotNull();
```
Comment (zoewangg) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AtomicCounterTest.java:139]: Can we add an assertion for the counter value?

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AtomicCounterExtension.java [hunk 2]
```diff
@@ -37,10 +39,11 @@
 import software.amazon.awssdk.enhanced.dynamodb.update.UpdateExpression;
 import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
 import software.amazon.awssdk.utils.CollectionUtils;
+import software.amazon.awssdk.utils.Logger;
 
 /**
- * This extension enables atomic counter attributes to be written to the database.
- * The extension is loaded by default when you instantiate a
+ * This extension enables atomic counter attributes to be changed in DynamoDb by creating instructions for modifying
```
Comment (zoewangg) [inline services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AtomicCounterExtension.java:45]: Can we update the Javadoc to mention that users updating the fields marked with atomic counter manually has no effect?

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AtomicCounterTest.java (hunk 1 above)
Comment (cenedhryn) [reply services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/AtomicCounterTest.java]: yes

**Review by zoewangg — APPROVED** [review]

File: services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AtomicCounterExtension.java (hunk 2 above)
Comment (cenedhryn) [reply services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/AtomicCounterExtension.java:45]: yes

**Pull request merged**